moshi/src/main/java/com/squareup/moshi/Moshi.kt

```diff
@@ -0,0 +1,388 @@
+        }
+        i++
+      }
+
+      // We might need to know about this cache key later in this call. Prepare for that.
+      val lookup = Lookup<Any>(type, fieldName, cacheKey)
+      callLookups += lookup
+      stack += lookup
+      return null
+    }
+
+    /** Sets the adapter result of the current lookup. */
+    fun <T> adapterFound(result: JsonAdapter<T>) {
+      @Suppress("UNCHECKED_CAST")
+      val currentLookup = stack.last() as Lookup<T>
+      currentLookup.adapter = result
+    }
+
+    /**
+     * Completes the current lookup by removing a stack frame.
+     *
+     * @param success true if the adapter cache should be populated if this is the topmost lookup.
+     */
+    fun pop(success: Boolean) {
+      stack.removeLast()
+      if (!stack.isEmpty()) return
+      lookupChainThreadLocal.remove()
+      if (success) {
+        synchronized(adapterCache) {
+          var i = 0
+          val size = callLookups.size
+          while (i < size) {
+            val lookup = callLookups[i]
+            val replaced = adapterCache.put(lookup.cacheKey, lookup.adapter)
+            if (replaced != null) {
+              @Suppress("UNCHECKED_CAST")
+              (lookup as Lookup<Any>).adapter = replaced as JsonAdapter<Any>
+              adapterCache[lookup.cacheKey] = replaced
+            }
+            i++
+          }
+        }
+      }
+    }
+
+    fun exceptionWithLookupStack(e: IllegalArgumentException): IllegalArgumentException {
+      // Don't add the lookup stack to more than one exception; the deepest is sufficient.
+      if (exceptionAnnotated) return e
+      exceptionAnnotated = true
+      val size = stack.size
+      if (size == 1 && stack.first().fieldName == null) return e
+      val errorMessage = buildString {
+        append(e.message)
+        for (lookup in stack.asReversed()) {
+          append("\nfor ").append(lookup.type)
+          if (lookup.fieldName != null) {
+            append(' ').append(lookup.fieldName)
+          }
+        }
+      }
+      return IllegalArgumentException(errorMessage, e)
+    }
+  }
+
+  /** This class implements `JsonAdapter` so it can be used as a stub for re-entrant calls. */
+  internal class Lookup<T>(val type: Type, val fieldName: String?, val cacheKey: Any) : JsonAdapter<T>() {
+    var adapter: JsonAdapter<T>? = null
+
+    override fun fromJson(reader: JsonReader): T? {
+      return withAdapter { fromJson(reader) }
+    }
+
+    override fun toJson(writer: JsonWriter, value: T?) {
+      @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS") // TODO remove after JsonAdapter is migrated
+      withAdapter { toJson(writer, value) }
+    }
+
+    private inline fun <R> withAdapter(body: JsonAdapter<T>.() -> R): R {
+      return checkNotNull(adapter) { "JsonAdapter isn't ready" }.body()
+    }
+
+    override fun toString(): String {
+      return adapter?.toString() ?: super.toString()
+    }
+  }
+
+  internal companion object {
+    @JvmField
```

Comment:
Can remove once all internal code is konverted?

Reply:
We could, though the only other place it's used is from MoshiTest and we won't be converting that to Kotlin right?